Release/v2.2.0 beta.3

robmoffat opened the pull request:
## Describe your change

<!--- Describe your change here-->

### Related Issue
<!--- This project prefers to accept pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- Please [link to the issue here](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue) by writing "resolves #123456" here: -->

## Contributor License Agreement

<!--- All contributions to FDC3 must be made under an active contributor license agreement and the [Community Specification License](https://github.com/finos/FDC3/blob/main/LICENSE.md). This will be checked by the EasyCLA tool (https://easycla.lfx.linuxfoundation.org/) that runs automatically on every PR. If you've not contributed to FDC3 before, look for a comment on your PR shortly after it is raised and follow the instructions to establish a CLA or have it acknowledged by the EasyCLA tool. -->

- [ ] I acknowledge that a contributor license agreement is required and that I have one in place or will seek to put one in place ASAP.

## Review Checklist

<!--- Checklist to be completed by reviewers, and pre-checked by the authors of a PR -->

- [ ] **Issue**: If a change was made to the FDC3 Standard, was an issue linked above?
- [ ] **CHANGELOG**: Is a *CHANGELOG.md* entry included?
- [ ] **API changes**: Does this PR include changes to any of the FDC3 APIs (`DesktopAgent`, `Channel`, `PrivateChannel`, `Listener`, `Bridging`)?
  - [ ] **Docs & Sources**: If yes, were both documentation (/docs) and sources updated?<br/>
        *JSDoc comments on interfaces and types should be matched to the main documentation in /docs*
  - [ ] **Conformance tests**: If yes, are conformance test definitions (/toolbox/fdc3-conformance) still correct and complete?<br/>
        *Conformance test definitions should cover all **required** aspects of an FDC3 Desktop Agent implementation, which are usually marked with a **MUST** keyword, and  optional features (**SHOULD** or **MAY**) where the format of those features is defined*
  - [ ] **Schemas**: If yes, were changes applied to the Bridging and FDC3 for Web protocol schemas?<br/>
        *The Web Connection protocol and Desktop Agent Communication Protocol schemas must be able to support all necessary aspects of the Desktop Agent API, while Bridging must support those aspects necessary for Desktop Agents to communicate with each other*
    - [ ] If yes, was code generation (`npm run build`) run and the results checked in?<br/>
        *Generated code will be found at `/src/api/BrowserTypes.ts` and/or `/src/bridging/BridgingTypes.ts`*
- [ ] **Context types**: Were new Context type schemas created or modified in this PR?
  - [ ] Were the [field type conventions](https://fdc3.finos.org/docs/context/spec#field-type-conventions) adhered to?
  - [ ] Was the `BaseContext` schema applied via `allOf` (as it is in existing types)?
  - [ ] Was a `title` and `description` provided for all properties defined in the schema?
  - [ ] Was at least one example provided?
  - [ ] Was code generation (`npm run build`) run and the results checked in?<br/>
        *Generated code will be found at `/src/context/ContextTypes.ts`*
- [ ] **Intents**: Were new Intents created in this PR?
  - [ ] Were the [intent name prefixes](https://fdc3.finos.org/docs/intents/spec#intent-name-prefixes) and other [naming conventions & characteristics](https://fdc3.finos.org/docs/intents/spec#naming-conventions) adhered to?
  - [ ] Was the new intent added to the list in the [Intents Overview](https://fdc3.finos.org/docs/intents/spec#standard-intents)?

bingenito commented:
@robmoffat Do the prettify changes to workbench need to be in this PR?

robmoffat commented:
Now I'm also seeing this: 

```
Inputs schema files: api/api.schema.json | api/common.schema.json | api/broadcastRequest.schema.json | api/findInstancesRequest.schema.json | api/findInstancesResponse.schema.json | api/findIntentRequest.schema.json | api/findIntentResponse.schema.json | api/findIntentsByContextRequest.schema.json | api/findIntentsByContextResponse.schema.json | api/getAppMetadataRequest.schema.json | api/getAppMetadataResponse.schema.json | api/openRequest.schema.json | api/openResponse.schema.json | api/raiseIntentRequest.schema.json | api/raiseIntentResponse.schema.json | api/raiseIntentResultResponse.schema.json | ../../fdc3-context/schemas/context/context.schema.json | bridging
Output file argument: ../generated/bridging/BridgingTypes.ts
command to run: ../../../node_modules/.bin/quicktype --prefer-const-values --prefer-unions -s schema -o ../generated/bridging/BridgingTypes.ts  -S api/api.schema.json -S api/common.schema.json -S api/broadcastRequest.schema.json -S api/findInstancesRequest.schema.json -S api/findInstancesResponse.schema.json -S api/findIntentRequest.schema.json -S api/findIntentResponse.schema.json -S api/findIntentsByContextRequest.schema.json -S api/findIntentsByContextResponse.schema.json -S api/getAppMetadataRequest.schema.json -S api/getAppMetadataResponse.schema.json -S api/openRequest.schema.json -S api/openResponse.schema.json -S api/raiseIntentRequest.schema.json -S api/raiseIntentResponse.schema.json -S api/raiseIntentResultResponse.schema.json -S ../../fdc3-context/schemas/context/context.schema.json  --src bridging/agentErrorResponse.schema.json --src bridging/agentRequest.schema.json --src bridging/agentResponse.schema.json --src bridging/bridgeErrorResponse.schema.json --src bridging/bridgeRequest.schema.json --src bridging/bridgeResponse.schema.json --src bridging/broadcastAgentRequest.schema.json --src bridging/broadcastBridgeRequest.schema.json --src bridging/common.schema.json --src bridging/connectionStep.schema.json --src bridging/connectionStep2Hello.schema.json --src bridging/connectionStep3Handshake.schema.json --src bridging/connectionStep4AuthenticationFailed.schema.json --src bridging/connectionStep6ConnectedAgentsUpdate.schema.json --src bridging/findInstancesAgentErrorResponse.schema.json --src bridging/findInstancesAgentRequest.schema.json --src bridging/findInstancesAgentResponse.schema.json --src bridging/findInstancesBridgeErrorResponse.schema.json --src bridging/findInstancesBridgeRequest.schema.json --src bridging/findInstancesBridgeResponse.schema.json --src bridging/findIntentAgentErrorResponse.schema.json --src bridging/findIntentAgentRequest.schema.json --src bridging/findIntentAgentResponse.schema.json --src bridging/findIntentBridgeErrorResponse.schema.json --src bridging/findIntentBridgeRequest.schema.json --src bridging/findIntentBridgeResponse.schema.json --src bridging/findIntentsByContextAgentErrorResponse.schema.json --src bridging/findIntentsByContextAgentRequest.schema.json --src bridging/findIntentsByContextAgentResponse.schema.json --src bridging/findIntentsByContextBridgeErrorResponse.schema.json --src bridging/findIntentsByContextBridgeRequest.schema.json --src bridging/findIntentsByContextBridgeResponse.schema.json --src bridging/getAppMetadataAgentErrorResponse.schema.json --src bridging/getAppMetadataAgentRequest.schema.json --src bridging/getAppMetadataAgentResponse.schema.json --src bridging/getAppMetadataBridgeErrorResponse.schema.json --src bridging/getAppMetadataBridgeRequest.schema.json --src bridging/getAppMetadataBridgeResponse.schema.json --src bridging/openAgentErrorResponse.schema.json --src bridging/openAgentRequest.schema.json --src bridging/openAgentResponse.schema.json --src bridging/openBridgeErrorResponse.schema.json --src bridging/openBridgeRequest.schema.json --src bridging/openBridgeResponse.schema.json --src bridging/privateChannelBroadcastAgentRequest.schema.json --src bridging/privateChannelBroadcastBridgeRequest.schema.json --src bridging/privateChannelEventListenerAddedAgentRequest.schema.json --src bridging/privateChannelEventListenerAddedBridgeRequest.schema.json --src bridging/privateChannelEventListenerRemovedAgentRequest.schema.json --src bridging/privateChannelEventListenerRemovedBridgeRequest.schema.json --src bridging/privateChannelOnAddContextListenerAgentRequest.schema.json --src bridging/privateChannelOnAddContextListenerBridgeRequest.schema.json --src bridging/privateChannelOnDisconnectAgentRequest.schema.json --src bridging/privateChannelOnDisconnectBridgeRequest.schema.json --src bridging/privateChannelOnUnsubscribeAgentRequest.schema.json --src bridging/privateChannelOnUnsubscribeBridgeRequest.schema.json --src bridging/raiseIntentAgentErrorResponse.schema.json --src bridging/raiseIntentAgentRequest.schema.json --src bridging/raiseIntentAgentResponse.schema.json --src bridging/raiseIntentBridgeErrorResponse.schema.json --src bridging/raiseIntentBridgeRequest.schema.json --src bridging/raiseIntentBridgeResponse.schema.json --src bridging/raiseIntentResultAgentErrorResponse.schema.json --src bridging/raiseIntentResultAgentResponse.schema.json --src bridging/raiseIntentResultBridgeErrorResponse.schema.json --src bridging/raiseIntentResultBridgeResponse.schema.json
Error: Key PrivateChannelEventListenerTypes not in schema object at https://fdc3.finos.org/schemas/next/api/common.schema.json#PrivateChannelEventListenerTypes.
```

@kriswest is this a filename length issue?

kriswest commented:
@robmoffat no that's a reference error of some sort that should not be occurring - possibly a merge error - but would be breaking the build (so no dist dir!).

Will need debug - i'll try and find some time to debug that (grumble, grumble)

kriswest commented:
There are couple of references in the bridging schemas to an element that no longer exists in the API schemas - it went away when we changed the event type names. I'll need to raise a PR to fix that.  I don't know why this hasn't been causing issues/been picked up before... possibly the generated files were hanging around and are now being ignored and it fails to generate? No, i'm sure bridging schema generation was working... I'll go check that on the branches we merged.

I also note the fdc3-workbench linting is broken, its ignoring pretty much every file as its misconfigured. Thats a problem for another day.

kriswest commented:
None of the 2.2.0 versions are correctly bundled, so theres not a lot to see in any of them, but the alphas do at least have a dist dir and /dist/src/index.js file (unlike beta.1). 

Hence, we need to look at fdc3-schema. There the alpha releases do contain generated sources (https://unpkg.com/browse/@finos/fdc3-schema@2.2.0-alpha.3/) while beta.1 does not (https://unpkg.com/browse/@finos/fdc3-schema@2.2.0-beta.1/). However, if I go back to alpha.3 on the fdc3-for-web-impl branch, bridging generation is also failing there. However, the old generated source file still exists and makes it into the build. I don't see the difference that would be causing it not to beta.1... A call to the `clean` script would do it as it deletes the generated files, however, I don't see one anywhere in the workflow and if I manually pack the module it does contain sources. Hence, its a bit of a mystery to me why this failed to build properly.

However, I am seeing the bridging schema error throughout, so that does need fixing still - I'm just not seeing how it resulted in a broken set of packages...

I had assumed this PR into main would be the culprit: https://github.com/finos/FDC3/pull/1459/files, it updated PrivateChannelEventListenerTypes in [schemas/api/common.schema.json](https://github.com/finos/FDC3/pull/1459/files#diff-20835d74047ef0849f2d6e7361901f53a920ca6dc9301db223385dca8a734f1b), but doesn't actually remove it. It actually gets removed (deuplicated with PrivateChannelEventTypes in api.schema.json) in https://github.com/finos/FDC3/pull/1466/files which was merged around the same time - but into fdc3-for-web-impl (I guess we were too far from the trunk).

Heres a PR that fixes the schema reference and hence code generation. Hopefully it also fixes the module build...

https://github.com/finos/FDC3/pull/1532/

robmoffat commented:
OK, I'm flummoxed.  I am trying to release using the existing package.yml.  The reason it worked last week was because we weren't actually doing a build.  Two publish operations occur, one to Github, one to NPM.  neither works.

First, I've added the token, which was not configured before.  This seems to now be set:

```
Run actions/setup-node@60edb5dd545a775178f52524783378180af0d1f8
  with:
    node-version: 20
    registry-url: https://registry.npmjs.org/
    always-auth: true
    token: ***
    check-latest: false
```

and 

```
Run actions/setup-node@60edb5dd545a775178f52524783378180af0d1f8
  with:
    node-version: 20
    registry-url: https://npm.pkg.github.com/
    always-auth: true
    token: ***
    check-latest: false
```


I've added a step to print out the whoami details, but this fails in both cases:

```
Run npm whoami --registry https://npm.pkg.github.com/
npm error code E403
npm error 403 403 Forbidden - GET https://npm.pkg.github.com/-/whoami - Permission denied
npm error 403 In most cases, you or one of your dependencies are requesting
npm error 403 a package version that is forbidden by your security policy, or
npm error 403 on a server you do not have access to.
npm error A complete log of this run can be found in: /home/runner/.npm/_logs/2025-03-06T17_06_39_242Z-debug-0.log
```

I've tried using a different key for the Github token, both as my robmoffat account and finos_admin. 

I don't know what else to try.  

@TheJuanAndOnly99 have you any ideas?

kriswest commented on .github/workflows/package.yml:
```diff
@@ -126,23 +126,41 @@ jobs:
+        always-auth: true
+        token: ${{ secrets[matrix.token-name] }}
+    
+    - name: Manually Configure npm Authentication
+      run: |
+        REGISTRY="${{ matrix.registry }}"
+        CLEAN_REGISTRY=${REGISTRY#https://}  # Strip https:// prefix
+        echo "//${CLEAN_REGISTRY}/:_authToken=${{ secrets[matrix.token-name] }}" > ~/.npmrc
+        echo "Generated .npmrc:"
+        cat ~/.npmrc  # Debug output
+        NODE_AUTH_TOKEN=${{ secrets[matrix.token-name] }}
```
Whats brought you to hear @robmoffat? In the old version the setup step on line 128 is setting the registry URL and line 150 the token (via an environment variable), which I know was working... We also did see beta.1 published to NPM successfully - I believe it was just the build that was broken in that case, right?

bingenito replied:
Yes, publishing was working fine, it was just empty package because of includes or directory structure issues.

robmoffat replied:
Separate problems I think.  What I'm doing above was trying to fix the npm install.  It fails before it even gets to publishing.  This didn't work either.

paulgoldsmith commented on .github/workflows/package.yml:
```diff
@@ -121,6 +121,12 @@ jobs:
       uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.7
 
+    - name: Install dependencies
+      run: npm install
```
Should this be `npm ci` (clean install)?

kriswest reviewed (approved):
LGTM - @paulgoldsmith's comment notwithstanding - `npm ci` looks like a sensible refinement to prevent unintended/untested dependency updates.

However, I would love to know what fixed the build/what the problem proved to be - and whether we really need all that manual config of a .npmrc file...

robmoffat commented:
> However, I would love to know what fixed the build/what the problem proved to be - and whether we really need all that manual config of a .npmrc file...

Pretty sure we don't, hence I've commented it out.  The problem seems to be that GitHub doesn't let you do an npm install from its own NPM repo.  Seems like this might be something that's changed.   The reason the npmrc stuff was in there was that I spent a lot of time  trying to understand that, as I was getting this error (see above comment):

```
npm error 403 403 Forbidden - GET https://npm.pkg.github.com/-/whoami - Permission denied
```

However, irrespective of the token I used and the way it was set up, it didn't work.  So, to get round this I reordered the build steps so that we do the `npm install` before changing to the GitHub NPM repo.   Publication to this repo is fine (which is why we were able to release beta.1).

julianna-ciq reviewed (approved):
LGTM, with a few caveats:
* I had a comment about documenting `getAgent` using `await`
* The PR shows jumps from v1 to v3, so I assume we still need to rebase off of main. But this branch is v2, so maybe v2 never went into main? I'm unclear what's happening with those numbers, but as long as someone knows why that happened, I don't consider it a blocker.

kriswest commented:
> However, irrespective of the token I used and the way it was set up, it didn't work. So, to get round this I reordered the build steps so that we do the `npm install` before changing to the GitHub NPM repo. Publication _to_ this repo is fine (which is why we were able to release beta.1).

Pretty sure that's how it worked in the past. I do NOT think it installed from the GitHub package repo.

robmoffat commented:
Can someone approve this?  Or let me know if I need to change something.  

@julianna-ciq - will rename the branch after merge.  I think there's UI for this in GitHub. 

thanks

kriswest reviewed (changes requested):
Thought I'd submitted this earlier in the week...

We've still got one dodgy readme + adding a couple of notes

kriswest commented on .github/workflows/package.yml:
```diff
@@ -111,38 +111,53 @@ jobs:
     - name: Report already published status
       if: steps.check-version.outputs.exists == '1' 
-      run: 'echo "package version already exists in the NPM registry"'
+      run: 'echo "package version already exists in the registry"'
```
Yeah you can see its hardcoded to the NPM registry here: 
https://github.com/tehpsalmist/npm-publish-status-action/blob/deb911186cfe5134094f49183364da10a986e4e7/src/main.ts#L39

kriswest commented on packages/fdc3-schema/README.md:
```diff
@@ -51,3 +51,8 @@ Finally, please note that the latest version of Quicktype (at the time of writin
   - A utility is provided in this repository (_../../quicktypeUtil.js_) to work around this bug by listing directories and constructing a call to quicktype where each file is provided as an individual source.
 - Setting the `--nice-property-names` TypeScript output option will crash the generator and can't be used.
 - The --debug setting must be the first argument passed or it will be ignored.
+
```
I'll see If I can sort this today

kriswest commented on .github/workflows/package.yml:
```diff
@@ -111,38 +111,53 @@ jobs:
     - name: Report already published status
       if: steps.check-version.outputs.exists == '1' 
-      run: 'echo "package version already exists in the NPM registry"'
+      run: 'echo "package version already exists in the registry"'
```
```suggestion
      run: 'echo "package version already exists in the *NPM* registry, skipping publishing"'
```

kriswest commented on .github/workflows/package.yml:
```diff
@@ -111,38 +111,53 @@ jobs:
     - name: Report not yet published status
       if: steps.check-version.outputs.exists == '0' 
-      run: 'echo "package version does not exist in the NPM registry, publishing..."'
+      run: 'echo "package version does not exist in the registry, publishing..."'
```
```suggestion
      run: 'echo "package version does not exist in the *NPM* registry, publishing..."'
```

kriswest reviewed (approved):
LGTM - will need another approver
